src/interrupts.rs

```diff
@@ -0,0 +1,305 @@
+}
+
+/// constants and structs needed for IOMMU. Grabbed from linux/vfio.h
+/// as this is a dynamically sized struct (has an array at the end) we need to use
+/// Dynamically Sized Types (DSTs) which can be found at
+/// https://doc.rust-lang.org/nomicon/exotic-sizes.html#dynamically-sized-types-dsts
+#[allow(non_camel_case_types)]
+#[repr(C)]
+struct vfio_irq_set<T: ?Sized> {
+    argsz: u32,
+    flags: u32,
+    index: u32,
+    start: u32,
+    count: u32,
+    data: T,
+}
+
+#[allow(non_camel_case_types)]
+#[repr(C)]
+struct vfio_irq_info {
+    argsz: u32,
+    flags: u32,
+    index: u32,		/* IRQ index */
+    count: u32,		/* Number of IRQs within this index */
+}
+
+const VFIO_IRQ_SET_DATA_NONE: u32 = (1 << 0); /* Data not present */
+const VFIO_IRQ_SET_DATA_EVENTFD: u32 = (1 << 2); /* Data is eventfd (s32) */
+const VFIO_IRQ_SET_ACTION_TRIGGER: u32 = (1 << 5); /* Trigger interrupt */
+const VFIO_DEVICE_GET_IRQ_INFO: u64 = 15213;
+const VFIO_DEVICE_SET_IRQS: u64 = 15214;
+pub const VFIO_PCI_MSI_IRQ_INDEX: u64 = 1;
+pub const VFIO_PCI_MSIX_IRQ_INDEX: u64 = 2;
+const VFIO_IRQ_INFO_EVENTFD: u32 = (1 << 0);
+
+impl Interrupts {
+    /// Setup VFIO interrupts by checking the `device_fd` for which interrupts this device supports.
+    /// Returns the supported interrupt type.
+    pub fn vfio_setup_interrupt(&mut self, device_fd: RawFd) -> Result<(), Box<dyn Error>> {
+        for index in (0..(VFIO_PCI_MSIX_IRQ_INDEX + 1)).rev() {
+            let irq_info: vfio_irq_info = vfio_irq_info {
+                argsz: mem::size_of::<vfio_irq_info>() as u32,
+                index: index as u32,
+                flags: 0,
+                count: 0,
+            };
+
+            if unsafe { libc::ioctl(device_fd, VFIO_DEVICE_GET_IRQ_INFO, &irq_info) } == -1 {
+                return Err(format!(
+                    "failed to VFIO_DEVICE_GET_IRQ_INFO for index {}. Errno: {}", index,
+                    unsafe { *libc::__errno_location() }
+                ).into());
+            }
+
+            if (irq_info.flags & VFIO_IRQ_INFO_EVENTFD) == 0 {
+                continue;
+            }
+
+            self.interrupt_type = index;
+            return Ok(());
+        }
+
+        self.interrupt_type = 0;
+        Ok(())
+    }
+}
+
+impl InterruptsQueue {
+
+    /// Add the `event_fd` file descriptor to epoll.
+    pub fn vfio_epoll_ctl(&mut self, event_fd: RawFd) -> Result<(), Box<dyn Error>> {
+        let event: Event = Event {
+            events: libc::EPOLLIN as u32,
+            data: event_fd as u64
+        };
+
+        let epoll_fd = epoll::create(false)?;
+
+        epoll::ctl(epoll_fd, epoll::ControlOptions::EPOLL_CTL_ADD, event_fd, event)?;
+
+        self.vfio_epoll_fd = epoll_fd;
+        Ok(())
+    }
+
+    /// Waits for events on the epoll instance referred to by the file descriptor `epoll_fd`.
+    /// The memory area pointed to by events will contain the events that will be available for the caller.
```

Comment:
Please insert a new line after the head line, e.g.

```rust
/// x does y.
///
/// This can be used to ...
fn x() {}
```

All in all, please adjust your doc comments to be more consistent with existing comments (only doc comments, the other comments are good to go)